TLS: Introduce *-alias to select specific key to use (or cert to trust)

famod opened the pull request:
Resolves #17424

Notes on store modifications in integration-tests/vertx-http:
- added key pair to server-keystore.jks with alias "anotherserver" (which comes before the existing "server" pair)
- added key pair to new client-keystore-2.jks and added its cert chain to server-truststore.jks with alias "mykey-2"
- renamed existing "mykey" to "mykey-1" in server-truststore.jks
- renamed client-keystore.jks to client-keystore-1.jks

---

More notes:
- I unified the creation of `KeyStoreOptions` in the recorder
- I went with the [parameter naming suggestion](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Any.20way.20to.20specify.20SSL.20key.20alias.20when.20using.20a.20keystore.20.3F/near/242353245) but maybe those names are a bit misleading because those are aliases for _entries in_ stores, not aliases _of_ stores, but then again I might be splitting hairs here

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/CertificateConfig.java:
```diff
@@ -75,7 +75,14 @@
     /**
-     * An optional trust store which holds the certificate information of the certificates to trust
+     * An optional parameter to select a specific key in the key store. If no such alias is defined and there are multiple keys
+     * in the store, behavior is undefined.
```
Probably worth mentioning SNI.

cescoffier replied:
I found it a bit dangerous to ignore the alias if one is set and none match. Anything we can do to reject the configuration in this case?

sberyozkin replied:
+1, if the user specifies the alias then it has to match or fail

sberyozkin replied:
I wonder if it should be detected at the Vert.x level though ? Perhaps it is already detected ?

famod replied:
I did test a non-matching alias and was surprised myself that vert.x is not bailing out with an error.
I do also think that this is not user-friendly but do we really want to handle this in Quarkus? AFAICS we would need to "open" the store ourself to check it.

sberyozkin replied:
@famod I agree it is not a Quarkus-level problem since Vert.x is managing the stores. Can you please open a Vert.x issue for the error be reported if no matching key exists ?
@cescoffier Do you agree ?

famod replied:
@sberyozkin
>  Can you please open a Vert.x issue for the error be reported if no matching key exists ?

I'll try. 😉 

@cescoffier
> Probably worth mentioning SNI.

I have only a vague understanding of SNI. Do you have a suggestion? Maybe something like:
> If no such alias is defined and there are multiple keys in the store, behavior is undefined (unless SNI is used).

cescoffier replied:
Yes something like this. SNI allows having multiple certificates and the server name allows using the right one. 
So something like: 

```
When SNI is disabled, if the key store contains multiple keys and no alias is specified, the behavior is undefined.
```

famod replied:
@cescoffier ok thanks, will use your suggestion.

I also suppose your are ok with having alias checking implemented in a later vertx release, as suggested by @sberyozkin.

cescoffier replied:
Yes.

famod commented:
I had a talk with my colleague @mickroll and I think this parameter naming question becomes more relevant:
In JBoss, Spring Boot etc. you can also specify a password for the key (not the store!) and I'm planning on requesting this in vert.x (and adding it to Quarkus once it has been released).
This would lead to a conflict with `key-store-password` if `key-store-alias` stays like this.

Maybe it would be better to have:
- `key-store-key-alias`
- `key-store-key-password`
?
Or even a "sub-section":
- `key-store-key.alias`
- `key-store-key.password`
?
Another approach would be to not use `key-store-` prefix at all for this...

WDYT?

sberyozkin commented:
@famod 
```
    key-store-key-alias
    key-store-key-password
```
are definitely better, and for this PR in particular, `key-store-key-alias` as it is exactly what it is, thanks

famod committed:
TLS: Introduce *-alias to select specific key to use (or cert to trust)

Resolves #17424

Notes on store modifications in integration-tests/vertx-http:
- added key pair to server-keystore.jks with alias "anotherserver" (which comes before the existing "server" pair)
- added key pair to new client-keystore-2.jks and added its cert chain to server-truststore.jks with alias "mykey-2"
- renamed existing "mykey" to "mykey-1" in server-truststore.jks
- renamed client-keystore.jks to client-keystore-1.jks

famod commented:
Remarks have been addressed. Btw, I went with `trust-store-cert-alias` for the trust store param.

sberyozkin commented:
Looks it is OK to merge, Clement @cescoffier has also approved, the failing test on Windows is not related to this PR
